cmd/k8s-nameserver/main.go

```diff
@@ -0,0 +1,211 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+
+//go:build !plan9
+
+package main
+
+// k8s-nameserver is a simple nameserver implementation meant to be used with
+// k8s-operator to allow to resolve magicDNS names of Tailscale services in a
+// Kubernetes cluster.
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"log"
+	"net"
+	"net/netip"
+	"os"
+
+	"github.com/fsnotify/fsnotify"
+	"tailscale.com/net/dns/resolver"
+	"tailscale.com/net/tsdial"
+	"tailscale.com/types/logger"
+	"tailscale.com/util/dnsname"
+)
+
+const (
+	defaultDNSConfigDir = "/config"
+	defaultDNSFile      = "dns.json"
+)
+
+func main() {
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	logger := log.Printf
+
+	res := resolver.New(logger, nil, nil, &tsdial.Dialer{Logf: logger})
+
+	var configReader configReaderFunc = func() ([]byte, error) {
+		return os.ReadFile(fmt.Sprintf("%s/%s", defaultDNSConfigDir, defaultDNSFile))
+	}
+
+	c := make(chan string)
+	watcher, err := fsnotify.NewWatcher()
+	if err != nil {
+		panic(err)
+	}
+	defer watcher.Close()
+	go func() {
+		logger("starting file watch for %s", defaultDNSConfigDir)
+		for {
+			select {
+			// TODO (irbekrm): it appears that we get a whole bunch
+			// of different events (except for the WRITE one that
+			// fsnotify recommends watching) on an update. We should
+			// batch updates together to avoid updating resolver
+			// config multiple times for each configmap update.
+			case event, ok := <-watcher.Events:
+				if !ok {
+					logger("watcher finished")
+					cancel()
+					return
+				}
+				// it seems like an update to the file has a
+				// potential to produce different types of
+				// events, but only one per configmap update.
+				// fsnotify docs suggest to only react to Write
+				// events, however we should be safe to react to
+				// other events too as we run in a container and
+				// there shouldn't be random changes to file
+				// metadata
+				msg := fmt.Sprintf("new file event: %v", event)
+				c <- msg
+
+			case err, ok := <-watcher.Errors:
+				if !ok {
+					logger("errors watcher finished: %v", err)
+					cancel()
+					return
+				}
+				if err != nil {
+					logger("error watching directory: %w", err)
+					cancel()
+					return
+				}
+			}
+		}
+	}()
+	err = watcher.Add(defaultDNSConfigDir)
+	if err != nil {
+		panic(err)
+	}
+
+	ns := &nameserver{
+		configReader:  configReader,
+		configWatcher: c,
+		logger:        logger,
+		res:           *res,
+	}
+
+	if err := ns.run(ctx, cancel); err != nil {
+		panic(fmt.Errorf("error running nameserver: %w", err))
+	}
+
+	addr, err := net.ResolveUDPAddr("udp", ":1053")
+	if err != nil {
+		panic("error resolving UDP address")
+	}
+	conn, err := net.ListenUDP("udp", addr)
+	if err != nil {
+		panic(fmt.Sprintf("error opening udp connection: %v", err))
+	}
+	go func() {
+		<-ctx.Done()
+		conn.Close()
+	}()
+
+	logger("k8s-nameserver listening on: %v", addr)
+
+	for {
```

Comment:
Do we really need to implement our own UDP server here and proxy requests to `tailscale.com/net/dns/resolver`? If we only wanrt to respond to a few DNS names based on a config file, it might be easier to just use miekg's dns.Server (https://jameshfisher.com/2017/08/04/golang-dns-server/)

What tailscale.com/net/dns/resolver also gives us is [automatic forwarding](https://github.com/tailscale/tailscale/blob/51d3220153e25f8144ca7186a0a5d851ffedd32b/net/dns/resolver/tsdns.go#L284) of DNS requests for names it does not keep track of, but we don't actually configure upstream resolvers to forward to, and I am not sure that forwarding unknown requests to another DNS server is actually desired here.

Reply:
I am not very strongly opinionated on this.
This uses the internal DNS resolver framework that we have to avoid pulling in another implementation - I originally wanted to use CoreDNS, but we had a chat with @maisem and came to an agreement that we should use one of the implementations we already have in codebase

> What tailscale.com/net/dns/resolver also gives us is [automatic forwarding](https://github.com/tailscale/tailscale/blob/51d3220153e25f8144ca7186a0a5d851ffedd32b/net/dns/resolver/tsdns.go#L284) of DNS requests for names it does not keep track of, but we don't actually configure upstream resolvers to forward to, and I am not sure that forwarding unknown requests to another DNS server is actually desired here.

In practice this should not be a problem- this nameserver will only be used as a [stub nameserver for ts.net subdomains]() and [I have added `ts.net`](https://github.com/tailscale/tailscale/pull/9201/commits/7ed55c26a2e52bb7e8cc41efa2b302817b33334d#diff-e0c39cc52dbd9188fc43ba19c1955045c74511bdcbbd9978230fdeb42714a712R212-R214) to [local domains](https://github.com/tailscale/tailscale/blob/v1.48.2/net/dns/resolver/tsdns.go#L74-L76) of the resolver config, so this should never receive any other requests than for ts.net and if it does not find a record for a ts.net subdomain, it should not attempt to forward to upstream resolver even if one was configured